### PR TITLE
Bugfix/separate test requirements

### PR DESCRIPTION
**Pull request by ChristopherGS**
@solegalli 

I noticed this error when installing feature-engine in another project:
image

Which made me realise that feature-engine setup.py has pytest down as a project requirement, which is not strictly true. I've therefore separated the test and library requirements. For another example of this, see the sanic framework: https://github.com/huge-success/sanic/blob/master/setup.py#L141

Once I made this change, I then went to run the tests and saw that they were a bit of date. I've updated them so they are all passing. In some cases, I did delete a few tests where it wasn't clear to me what to update. I figure it's better to have some working than none at all, but please make sure I haven't removed something important.

### CODE REVIEW

File: feature_engine/missing_data_imputers.py [hunk 1]
```diff
@@ -196,7 +196,7 @@ def fit(self, X, y=None, ):
         
         elif self.tail == 'left':
             if self.distribution == 'gaussian':
-                self.imputer_dict_ = (X[self.variables].mean()-self.fold*X[self.variable_].std()).to_dict()
+                self.imputer_dict_ = (X[self.variables].mean()-self.fold*X[self.variables].std()).to_dict()
```
Comment (ChristopherGS) [inline feature_engine/missing_data_imputers.py]: I think this is an important bugfix

File: feature_engine/tests/test_categorical_encoder.py [hunk 2]
```diff
@@ -2,192 +2,10 @@
 # License: BSD 3 clause
 
 import pandas as pd
-import pytest
 
-from categorical_encoder import CategoricalEncoder, RareLabelEncoder
+from feature_engine.categorical_encoders import RareLabelCategoricalEncoder
 
 
-def test_CategoricalEncoder_count():
```
Comment (ChristopherGS) [inline feature_engine/tests/test_categorical_encoder.py]: these have all been renamed with (I think) different implementations so I removed them

File: test_requirements.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+pytest>=4.3.1,<6.0
```
Comment (ChristopherGS) [inline test_requirements.txt]: I tested the tests pass with v4.31 and the latest v.5.2.4

File: feature_engine/missing_data_imputers.py (hunk 1 above)
Comment (solegalli) [reply feature_engine/missing_data_imputers.py]: Massive bugfix, thanks!

File: feature_engine/tests/test_categorical_encoder.py [hunk 4]
```diff
@@ -2,192 +2,10 @@
...
-    df = pd.DataFrame(df)
-    
-    transf_df = {'category': [0.25,0.25,0.25,0.25,0.25,
-                              0.25,0.25,0.25,0.25,0.25,
-                              0.50,0.50,0.50,0.50,0.50,
-                              0.50,1.00,1.00,1.00,1.00],
-                 'target' : [1,1,0,0,0,0,0,0,0,0,1,1,0,0,0,0, 1,1,0,0]}
-    transf_df = pd.DataFrame(transf_df)
-    
-    encoder = CategoricalEncoder(encoding_method='ratio')
```
Comment (solegalli) [inline feature_engine/tests/test_categorical_encoder.py]: replace by:
encoder = WoERatioCategoricalEncoder(encoding_method='ratio', variables = ['category'])

File: feature_engine/tests/test_categorical_encoder.py [hunk 5]
```diff
@@ -2,192 +2,10 @@
...
-    encoder.fit(df['category'].to_frame(), df['target'] )
-    X = encoder.transform(df['category'].to_frame())
-    
-    pd.testing.assert_frame_equal(X, transf_df['category'].to_frame())
-    assert encoder.variables_ == ['category']
```
Comment (solegalli) [inline feature_engine/tests/test_categorical_encoder.py]: assert encoder.variables == ['category']

File: feature_engine/tests/test_categorical_encoder.py [hunk 6]
```diff
@@ -2,192 +2,10 @@
...
-    df = pd.DataFrame(df)
-    
-    transf_df = {'category': [0.25,0.25,0.25,0.25,0.25,
-                              0.25,0.25,0.25,0.25,0.25,
-                              0.50,0.50,0.50,0.50,0.50,
-                              0.50,10000.00,10000.00,10000.00,10000.00],
-                 'target' : [1,1,0,0,0,0,0,0,0,0,1,1,0,0,0,0, 1,1,0,0]}
-    transf_df = pd.DataFrame(transf_df)
-    
-    encoder = CategoricalEncoder(encoding_method='ratio')
```
Comment (solegalli) [inline feature_engine/tests/test_categorical_encoder.py]: encoder = WoERatioCategoricalEncoder(encoding_method='ratio', variables = ['category'])

File: feature_engine/tests/test_categorical_encoder.py [hunk 7]
```diff
@@ -2,192 +2,10 @@
...
-    df = pd.DataFrame(df)
-    
-    transf_df = {'category': [-1.386294,-1.386294,-1.386294,-1.386294,-1.386294,
-                              -1.386294,-1.386294,-1.386294,-1.386294,-1.386294,
-                              0.693147,0.693147,0.693147,0.693147,0.693147,
-                              0.693147,0.000000,0.000000,0.000000,0.000000],
-                 'target' : [1,1,0,0,0,0,0,0,0,0,1,1,0,0,0,0, 1,1,0,0]}
-    transf_df = pd.DataFrame(transf_df)
-    
-    encoder = CategoricalEncoder(encoding_method='woe')
```
Comment (solegalli) [inline feature_engine/tests/test_categorical_encoder.py]: encoder = WoERatioCategoricalEncoder(encoding_method='woe', variables = ['category'])

File: feature_engine/tests/test_categorical_encoder.py [hunk 8]
```diff
@@ -2,192 +2,10 @@
...
-    encoder.fit(df['category'].to_frame(), df['target'] )
-    X = encoder.transform(df['category'].to_frame())
-    
-    pd.testing.assert_frame_equal(X, transf_df['category'].to_frame())
-    assert encoder.variables_ == ['category']
```
Comment (solegalli) [inline feature_engine/tests/test_categorical_encoder.py]: assert encoder.variables == ['category']

File: feature_engine/tests/test_categorical_encoder.py [hunk 9]
```diff
@@ -2,192 +2,10 @@
...
-    df = pd.DataFrame(df)
-    
-    transf_df = {'category': [-1.386294,-1.386294,-1.386294,-1.386294,-1.386294,
-                              -1.386294,-1.386294,-1.386294,-1.386294,-1.386294,
-                              0.693147,0.693147,0.693147,0.693147,0.693147,
-                              0.693147,9.210340,9.210340,9.210340,9.210340],
-                 'target' : [1,1,0,0,0,0,0,0,0,0,1,1,0,0,0,0, 1,1,0,0]}
-    transf_df = pd.DataFrame(transf_df)
-    
-    encoder = CategoricalEncoder(encoding_method='woe')
```
Comment (solegalli) [inline feature_engine/tests/test_categorical_encoder.py]: encoder = WoERatioCategoricalEncoder(encoding_method='woe', variables = ['category'])

File: feature_engine/tests/test_categorical_encoder.py [hunk 10]
```diff
@@ -2,192 +2,10 @@
...
-    df = pd.DataFrame(df)
-    
-    transf_df = {'category': [-1.386294,-1.386294,-1.386294,-1.386294,-1.386294,
-                              -1.386294,-1.386294,-1.386294,-1.386294,-1.386294,
-                              0.693147,0.693147,0.693147,0.693147,0.693147,
-                              0.693147,-9.210340,-9.210340,-9.210340,-9.210340],
-                 'target' : [1,1,0,0,0,0,0,0,0,0,1,1,0,0,0,0, 1,1,0,0]}
-    transf_df = pd.DataFrame(transf_df)
-    
-    encoder = CategoricalEncoder(encoding_method='woe')
```
Comment (solegalli) [inline feature_engine/tests/test_categorical_encoder.py]: encoder = WoERatioCategoricalEncoder(encoding_method='woe', variables = ['category'])

File: feature_engine/tests/test_missing_data_imputer.py [hunk 11]
```diff
@@ -1,160 +1,111 @@
...
-from missing_data_imputation import RandomSampleImputer
-from missing_data_imputation import EndTailImputer
-from missing_data_imputation import na_capturer
-from missing_data_imputation import CategoricalImputer
-from missing_data_imputation import ArbitraryImputer
+from feature_engine.missing_data_imputers import MeanMedianImputer
+from feature_engine.missing_data_imputers import RandomSampleImputer
+from feature_engine.missing_data_imputers import EndTailImputer
+from feature_engine.missing_data_imputers import CategoricalVariableImputer
+
```
Comment (solegalli) [inline feature_engine/tests/test_missing_data_imputer.py]: add:
from feature_engine.missing_data_imputers import ArbitraryNumberImputer
from feature_engine.missing_data_imputers import FrequentCategoryImputer
from feature_engine.missing_data_imputers import AddNaNBinaryImputer

File: feature_engine/tests/test_missing_data_imputer.py [hunk 12]
```diff
@@ -1,160 +1,111 @@
...
 def test_MeanMedianImputer():
-    imputer = MeanMedianImputer(imputation_method='median')
-    imputer.fit(df_test, variables = ['Age'])
+    imputer = MeanMedianImputer(imputation_method='median', variables=['Age'])
+    imputer.fit(df_test)
     X = imputer.transform(df_test)
     df_transf = df_test.copy()
     df_transf['Age'] = df_test['Age'].fillna(31.5)
     pd.testing.assert_frame_equal(X, df_transf)
-    assert imputer.variables_ == ['Age']
```
Comment (solegalli) [inline feature_engine/tests/test_missing_data_imputer.py]: why did you remove this one? can we not keep:
assert imputer.variables == ['Age']

File: feature_engine/tests/test_missing_data_imputer.py [hunk 13]
```diff
@@ -1,160 +1,111 @@
...
     
-    imputer = MeanMedianImputer(imputation_method='mean')
-    imputer.fit(df_test, variables = ['Age'])
+    imputer = MeanMedianImputer(imputation_method='mean', variables=['Age'])
+    imputer.fit(df_test)
     X = imputer.transform(df_test)
     df_transf = df_test.copy()
     df_transf['Age'] = df_test['Age'].fillna(29.946315789473687)
     pd.testing.assert_frame_equal(X, df_transf)
-    assert imputer.variables_ == ['Age']
```
Comment (solegalli) [inline feature_engine/tests/test_missing_data_imputer.py]: assert imputer.variables == ['Age']

File: feature_engine/tests/test_missing_data_imputer.py [hunk 14]
```diff
@@ -1,160 +1,111 @@
...
 def test_EndTailImputer():
-    imputer = EndTailImputer(distribution='gaussian', tail='right', fold=3)
-    imputer.fit(df_test, variables = ['Age'])
+    imputer = EndTailImputer(distribution='gaussian', tail='right', fold=3, variables=['Age'])
+    imputer.fit(df_test)
     X = imputer.transform(df_test)
     df_transf = df_test.copy()
     df_transf['Age'] = df_test['Age'].fillna(73.98313089014289)
     pd.testing.assert_frame_equal(X, df_transf)
-    assert imputer.variables_ == ['Age']
```
Comment (solegalli) [inline feature_engine/tests/test_missing_data_imputer.py]: assert imputer.variables == ['Age']

File: feature_engine/tests/test_missing_data_imputer.py [hunk 15]
```diff
@@ -1,160 +1,111 @@
...
     
-    imputer = EndTailImputer(distribution='gaussian', tail='right', fold=1.5)
-    imputer.fit(df_test, variables = ['Age'])
+    imputer = EndTailImputer(distribution='gaussian', tail='right', fold=1.5, variables=['Age'])
+    imputer.fit(df_test)
     X = imputer.transform(df_test)
     df_transf = df_test.copy()
     df_transf['Age'] = df_test['Age'].fillna(51.96472333980829)
     pd.testing.assert_frame_equal(X, df_transf)
-    assert imputer.variables_ == ['Age']
```
Comment (solegalli) [inline feature_engine/tests/test_missing_data_imputer.py]: assert imputer.variables == ['Age']

File: feature_engine/tests/test_missing_data_imputer.py [hunk 16]
```diff
@@ -1,160 +1,111 @@
...
     
-    imputer = EndTailImputer(distribution='gaussian', tail='left', fold=2)
-    imputer.fit(df_test, variables = ['Age'])
+    imputer = EndTailImputer(distribution='gaussian', tail='left', fold=2, variables=['Age'])
+    imputer.fit(df_test)
     X = imputer.transform(df_test)
     df_transf = df_test.copy()
     df_transf['Age'] = df_test['Age'].fillna(0.5884390556942236)
     pd.testing.assert_frame_equal(X, df_transf)
-    assert imputer.variables_ == ['Age']
```
Comment (solegalli) [inline feature_engine/tests/test_missing_data_imputer.py]: assert imputer.variables == ['Age']

File: feature_engine/tests/test_missing_data_imputer.py [hunk 17]
```diff
@@ -1,160 +1,111 @@
...
     
-    imputer = EndTailImputer(distribution='skewed', tail='right', fold=3)
-    imputer.fit(df_test, variables = ['Age'])
+    imputer = EndTailImputer(distribution='skewed', tail='right', fold=3, variables=['Age'])
+    imputer.fit(df_test)
     X = imputer.transform(df_test)
     df_transf = df_test.copy()
     df_transf['Age'] = df_test['Age'].fillna(95.25)
     pd.testing.assert_frame_equal(X, df_transf)
-    assert imputer.variables_ == ['Age']
```
Comment (solegalli) [inline feature_engine/tests/test_missing_data_imputer.py]: assert imputer.variables == ['Age']

File: feature_engine/tests/test_missing_data_imputer.py [hunk 18]
```diff
@@ -1,160 +1,111 @@
...
     df_transf['Age'] = df_test['Age'].fillna(-36.0)
     pd.testing.assert_frame_equal(X, df_transf)
-    assert imputer.variables_ == ['Age']
     assert imputer.input_shape_ == (100,3)
     assert imputer.imputer_dict_ == {'Age':-36.0}
     assert X['Age'].isnull().sum() == 0
     
     
-def test_na_capturer():
-    imputer = na_capturer(tol=0.05)
```
Comment (solegalli) [inline feature_engine/tests/test_missing_data_imputer.py]: imputer = AddNaNBinaryImputer(variables=['Age', 'Cabin'])

File: feature_engine/tests/test_missing_data_imputer.py [hunk 19]
```diff
@@ -1,160 +1,111 @@
...
     df_transf = df_test.copy()
     df_transf['Age'] = df_test['Age'].fillna(-36.0)
     pd.testing.assert_frame_equal(X, df_transf)
-    assert imputer.variables_ == ['Age']
     assert imputer.input_shape_ == (100,3)
     assert imputer.imputer_dict_ == {'Age':-36.0}
     assert X['Age'].isnull().sum() == 0
     
     
-def test_na_capturer():
```
Comment (solegalli) [inline feature_engine/tests/test_missing_data_imputer.py]: rename to:
def test_missing_indicator():

File: feature_engine/tests/test_missing_data_imputer.py [hunk 20]
```diff
@@ -1,160 +1,111 @@
...
-    imputer = na_capturer(tol=0.05)
-    imputer.fit(df_test)
-    X = imputer.transform(df_test)
-    df_transf = df_test.copy()
-    df_transf['Age_na'] = np.where(df_test['Age'].isnull(),1,0)
-    df_transf['Cabin_na'] = np.where(df_test['Cabin'].isnull(),1,0)
-    pd.testing.assert_frame_equal(X, df_transf) 
-    assert imputer.variables_ == ['Age', 'Cabin']
```
Comment (solegalli) [inline feature_engine/tests/test_missing_data_imputer.py]: assert imputer.variables == ['Age', 'Cabin']

File: feature_engine/tests/test_missing_data_imputer.py [hunk 21]
```diff
@@ -1,160 +1,111 @@
...
-    assert imputer.input_shape_ == (100,3)
-    assert X.shape == (100,5)
```
Comment (solegalli) [inline feature_engine/tests/test_missing_data_imputer.py]: Not too sure this test will pass, but I can't fix it later if it doesn't.

File: feature_engine/tests/test_missing_data_imputer.py [hunk 22]
```diff
@@ -1,160 +1,111 @@
...
+    pd.testing.assert_frame_equal(X, df_transf)
     assert imputer.input_shape_ == (100,3)
-    assert imputer.imputer_dict_['Cabin'] == 'Missing'
-    assert imputer.imputer_dict_['Embarked'] == 'S'
     assert np.sum(X[['Cabin', 'Embarked']].isnull().sum()) == 0
-    
-    
-def test_ArbitraryImputer():
-    imputer = ArbitraryImputer()
-    imputer.fit(df_test, imputation_dictionary = {'Age':100,
```
Comment (solegalli) [inline feature_engine/tests/test_missing_data_imputer.py]: imputer.fit(df)

File: feature_engine/tests/test_missing_data_imputer.py [hunk 23]
```diff
@@ -1,160 +1,111 @@
...
-                                                  'Cabin':'Missing',
-                                                  'Embarked':'Unknown'})
-    X = imputer.transform(df_test)
-    df_transf = df_test.copy()
-    df_transf['Age'] = df_test['Age'].fillna(100)
```
Comment (solegalli) [inline feature_engine/tests/test_missing_data_imputer.py]: df_transf['Age'] = df_test['Age'].fillna(-999)

File: feature_engine/tests/test_missing_data_imputer.py [hunk 24]
```diff
@@ -1,160 +1,111 @@
...
-    df_transf['Cabin'] = df_test['Cabin'].fillna('Missing')
```
Comment (solegalli) [inline feature_engine/tests/test_missing_data_imputer.py]: delete

File: feature_engine/tests/test_missing_data_imputer.py [hunk 25]
```diff
@@ -1,160 +1,111 @@
...
-    df_transf['Embarked'] = df_test['Embarked'].fillna('Unknown')
```
Comment (solegalli) [inline feature_engine/tests/test_missing_data_imputer.py]: delete

File: feature_engine/tests/test_missing_data_imputer.py [hunk 26]
```diff
@@ -1,160 +1,111 @@
...
-    pd.testing.assert_frame_equal(X, df_transf) 
-    assert imputer.input_shape_ == (100,3)
-    assert imputer.imputer_dict_ == {'Age':100,'Cabin':'Missing','Embarked':'Unknown'}
```
Comment (solegalli) [inline feature_engine/tests/test_missing_data_imputer.py]: assert imputerarbitrary_number == -999

File: feature_engine/tests/test_missing_data_imputer.py [hunk 27]
```diff
@@ -1,160 +1,111 @@
...
-    assert np.sum(X.isnull().sum()) == 0
```
Comment (solegalli) [inline feature_engine/tests/test_missing_data_imputer.py]: assert np.sum(X['Age'].isnull().sum()) == 0
assert np.sum(X[imputer.variables].isnull().sum().sum()) == 0

File: test_requirements.txt (hunk 3 above)
Comment (solegalli) [reply test_requirements.txt]: @ChristopherGS I don't have enough words to thank you!!!!
You are amazing. Thank you for so much work.
